service/handlers.go

```diff
@@ -0,0 +1,185 @@
+			Message:      "Successfully updated domain name",
+			ResponseCode: http.StatusOK,
+			Error:        errResp,
+		}
+		js, err := json.Marshal(response)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		_, _ = w.Write(js)
+	}
+}
+
+func (h *BaseHandler) BouncedHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		errResp, domainName := getDomain(r)
+
+		err := h.updateOrCreate(domainName, 0, 1)
+		if err != nil {
+			errResp = true
+			fmt.Println("Error occurred in BounceHandler", err)
+		}
+
+		response := models.Response{
+			Message:      "Successfully updated domain name",
+			ResponseCode: http.StatusOK,
+			Error:        errResp,
+		}
+		js, err := json.Marshal(response)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		_, _ = w.Write(js)
+	}
+
+}
+
+func (h *BaseHandler) GetDomainHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		errResp, domainName := getDomain(r)
+
+		event, err := h.get(domainName)
+		if err != nil {
+			errResp = true
+			fmt.Println("Error occurred in GetDomainHandler", err)
+		}
+		domainType := determineDomain(event)
+
+		fmt.Println(event, domainType)
+		response := models.GetResponse{
+			Response: models.Response{
+				Message:      "Successfully updated domain name",
+				ResponseCode: http.StatusOK,
+				Error:        errResp,
+			},
+			Event:      event,
+			DomainType: domainType,
+		}
+
+		js, err := json.Marshal(response)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		_, _ = w.Write(js)
+	}
+}
+
+func getDomain(r *http.Request) (bool, string) {
+	// get URL parameters
+	vars := mux.Vars(r)
+	errResp := false
+	// find the metric name
+	domainName := vars["domainName"]
+
+	fmt.Println("msg", "Domain name received", "name", domainName)
+
+	return errResp, domainName
+}
+
+func determineDomain(event models.Event) string {
+	domainType := "unknown"
+	//receives more than 1000 “delivered” events
+	if event.Delivered > DomainThreshold {
+		domainType = "catch-all"
+	}
+	//A domain name is not a catch-all when it receives at least 1 “bounced” event
+	if event.Bounced >= BouncedThreshold {
+		domainType = "not catch-all"
+	}
+	return domainType
+}
+
+// These could be abstracted into there own DB layer
+func (h *BaseHandler) updateOrCreate(domain string, deliveredIncrease int64, bouncedIncrease int64) error {
+
+	var exists bool
+	var delivered int64
+	var bounced int64
+
+	err := h.db.QueryRow(context.Background(), "select exists(select 1 from events where domain=$1)", domain).Scan(&exists)
+	if err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "QueryRow exists failed: %v\n", err)
+	}
+	//does the entry exist?
+	if exists {
```

Comment:
Could an upsert work here?

Reply:
Upsert would be more efficient here yes. Thank you